Userland/Libraries/LibGfx/tiff_generator.py

```diff
@@ -0,0 +1,321 @@
+#include <AK/Variant.h>
+#include <AK/Vector.h>
+#include <LibGfx/Size.h>
+
+namespace Gfx {{
+
+class Metadata;
+
+namespace TIFF {{
+
+{export_enum_to_cpp(TIFFType, 'Type')}
+
+template<OneOf<u32, i32> x32>
+struct Rational {{
+    using Type = x32;
+    x32 numerator;
+    x32 denominator;
+}};
+
+{export_promoter()}
+
+// Note that u16 is not include on purpose
+using Value = Variant<u8, String, u32, Rational<u32>, i32, Rational<i32>>;
+
+// This enum is progressively defined across sections but summarized in:
+// Appendix A: TIFF Tags Sorted by Number
+{export_enum_to_cpp(Compression)}
+
+{export_enum_to_cpp(Predictor)}
+
+{HANDLE_TAG_SIGNATURE};
+
+}}
+
+"""
+
+    output += generate_metadata_class(tags)
+
+    output += '\n}\n'
+
+    return output
+
+
+def generate_tag_handler(tag: Tag) -> str:
+    not_in_type_list = f"({' && '.join([f'type != Type::{t.name}' for t in tag.types])})"
+
+    not_in_count_list = ''
+    if len(tag.counts) != 0:
+        not_in_count_list = f"|| ({' && '.join([f'count != {c}' for c in tag.counts])})"
+    pre_condition = fR"""if ({not_in_type_list}
+            {not_in_count_list})
+            return Error::from_string_literal("TIFFImageDecoderPlugin: Tag {tag.name} invalid");"""
+
+    check_value = ''
+    if tag.associated_enum is not None:
+        not_in_value_list = f"({' && '.join([f'v != {v.value}' for v in tag.associated_enum])})"
+        check_value = fR"""TRY(value[0].visit(
+            []({tiff_type_to_cpp(tag.types[0])} const& v) -> ErrorOr<void> {{
+                if ({not_in_value_list})
+                    return Error::from_string_literal("TIFFImageDecoderPlugin: Invalid value for tag {tag.name}");
+                return {{}};
+            }},
+            [&](auto const&) -> ErrorOr<void> {{
+                VERIFY_NOT_REACHED();
+            }}));
+"""
+
+    output = fR"""    case {tag.id}:
+        // {tag.name}
+        {pre_condition}
+        {check_value}
+        metadata.add_entry("{tag.name}"sv, move(value));
+        break;
+"""
+
+    return output
+
+
+def generate_tag_handler_file(tags: List[Tag]) -> str:
+    output = fR"""{LICENSE}
+
+#include <AK/Debug.h>
+#include <AK/String.h>
+#include <LibGfx/ImageFormats/TIFFMetadata.h>
+
+namespace Gfx::TIFF {{
+
+{HANDLE_TAG_SIGNATURE}
+{{
+    switch (tag) {{
+"""
+
+    output += '\n'.join([generate_tag_handler(t) for t in tags])
+
+    output += R"""
+    default:
+        dbgln_if(TIFF_DEBUG, "Unknown tag: {}", tag);
+    }
+
+    return {};
+}
+
+}
+"""
+    return output
+
+
+def main():
+    parser = argparse.ArgumentParser()
+    parser.add_argument('-o', '--output')
+    args = parser.parse_args()
+
+    output_path = Path(args.output)
+
+    with open(output_path / 'TIFFMetadata.h', 'w') as metadata_file:
+        metadata_file.write(generate_metadata_file(known_tags))
+
+    with open(output_path / 'TIFFTagHandler.cpp', 'w') as handler_file:
+        handler_file.write(generate_tag_handler_file(known_tags))
```

Comment:
this is where the "only write if changed" optimization would come in. Read the file, compare it to the generated string, and only update the file if the contents are different. this will keep the mtime of the file from causing unecessary rebuilds without needing extra cmake processes to do the work.